primitives/runtime/src/offchain/rpc.rs

```diff
@@ -0,0 +1,353 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A high-level helpers for making RPC calls from Offchain Workers.
+//!
+//! Example:
+//! ```rust,no_run
+//! use sp_runtime::offchain::rpc::{Request, Error};
+//! use sp_runtime::offchain::http;
+//! use sp_std::{vec::Vec, str};
+//!
+//! let request = Request {
+//! 	jsonrpc: "2.0",
+//! 	id: 1,
+//! 	method: "chain_getFinalizedHead",
+//! 	params: Vec::new(),
+//! 	timeout: 3_000,
+//! 	url: "http://localhost:9933"
+//! };
+//!
+//! let rpc_response = request.send();
+//!
+//! match rpc_response {
+//! 	Ok(response) => {
+//! 		if !response.result.is_empty() {
+//! 			log::info!("Rpc call result: {:?}", str::from_utf8(&response.result).unwrap());
+//! 		} else {
+//! 			log::error!("Rpc call error: code: {:?}, message: {:?}",
+//! 							response.error.code,
+//! 							str::from_utf8(&response.error.message).unwrap()
+//! 						);
+//! 		}
+//! 	},
+//! 	Err(e) => {
+//! 		log::error!("Rpc call error: {:?}", e);
+//! 		assert_eq!(Error::Http(http::Error::DeadlineReached), e);
+//! 	}
+//! }
+//! ```
+
+use serde::{Deserialize, Deserializer};
+use crate::{RuntimeDebug, offchain::Duration};
+use serde_json::json;
+use sp_std::{vec::Vec, str};
+use super::http;
+
+/// A rpc call error
+#[derive(Clone, PartialEq, Eq, RuntimeDebug)]
+pub enum Error {
+	/// Response can not be deserialized
+	Deserializing,
+	/// Http error
+	Http(http::Error),
+}
+
+/// A rpc result type
+pub type RpcResult = Result<Response, Error>;
+
+/// A rpc call error field
+#[derive(Deserialize, Default, RuntimeDebug)]
+pub struct RpcError {
+	/// error code
+	pub code: i32,
+	/// error message
+	#[serde(deserialize_with = "de_string_to_bytes")]
+	pub message: Vec<u8>,
+}
+
+/// A rpc call response
+#[derive(Deserialize, Default, RuntimeDebug)]
+pub struct Response {
+	/// rpc response jsonrpc
+	#[serde(deserialize_with = "de_string_to_bytes")]
+	pub jsonrpc: Vec<u8>,
+	/// rpc response result
+	#[serde(default)]
+	#[serde(deserialize_with = "de_string_to_bytes")]
+	pub result: Vec<u8>,
+	/// rpc response error struct
+	#[serde(default)]
+	pub error: RpcError,
+	/// rpc response id
+	pub id: u32,
+}
+
+/// A rpc call request
+pub struct Request<'a> {
+	/// rpc request jsonrpc
+	pub jsonrpc: &'a str,
+	/// rpc request id
+	pub id: u32,
+	/// rpc request method
+	pub method: &'a str,
+	/// rpc request params
+	pub params: Vec<&'a str>,
+	/// rpc request timeout
+	pub timeout: u64,
+	/// rpc request url
+	pub url: &'a str,
```

Comment:
Given the context of #6742 I'd avoid adding `url` here, so that the `Request` object can be re-used to send "internal" rpc requests.

Maybe let's move it to `send()`?

Reply:
We can also consider moving `timeout` so that this is kept as close to jsonrpc spec as possible.

Reply:
I follow up on this topic in another comment thread